### ISSUE TITLE
Quantum Gradient Descent implementation for TensorFlow

### ISSUE THREAD

**Issue by jackaraz** [PennyLaneAI/pennylane#2226]
### Feature details

Would dev-team be interested in implementing [Quantum gradient descent](https://arxiv.org/pdf/1612.01789.pdf) optimization for the TensorFlow backend?

### Implementation

An implementation of QGD optimizer based on Keras can be found in [here](https://gitlab.com/jackaraz/tnqcircuits/-/blob/main/src/tnqcircuits/optimizers/quantum_grad_descent.py) but this implementation requires a custom model which includes its own `train_on_batch` function (an example can be found [here](https://gitlab.com/jackaraz/tnqcircuits/-/blob/main/src/tnqcircuits/models/quantum_model.py#L72)) due to the reasons discussed in issue #2040 . Hence it limits the flexibility of the method for instance model can not be used in a hybrid environment which requires further implementations (example can be found in [here](https://gitlab.com/jackaraz/tnqcircuits/-/blob/main/src/tnqcircuits/models/hybrid_model.py)). All of these implementations are based on pennylane.

### How important would you say this feature is?

1: Not important. Would be nice to have.

**Comment by josh146** [comment]
Hi @jackaraz --- I was just looking through your [`tnqcircuits`](https://gitlab.com/jackaraz/tnqcircuits/) repo, a really nice package! Cool to also see that you have a TF compatible QNG optimizer 🙂

Adding TensorFlow specific optimization methods has been on our todo list for a long time, but an unanswered question was how to implement them - at the low-level gradient tape, or as a Keras optimizer? So it has mainly been an issue of design. Since you have experience here, any insight you have would be appreciated.

Regarding the quantum gradient descent optimizer specifically, I'm curious about your implementation, as the original paper took very much a fault-tolerant/non-NISQy approach?

**Comment by jackaraz** [comment]
Thanks @josh146 :)

> Adding TensorFlow specific optimization methods has been on our todo list for a long time, but an unanswered question was how to implement them - at the low-level gradient tape, or as a Keras optimizer? So it has mainly been an issue of design. Since you have experience here, any insight you have would be appreciated.

If the main goal is flexibility in terms of the user interface, I believe low-level implementation would be preferable because in our implementation ([based on this study](https://arxiv.org/abs/2202.10471)) user needs to know what they are doing; hence it's not a very "Keras-like" approach. This is mainly due to the metric tensor calculation, as we discussed in #2040 , TF is not able to remember the trainable parameters outside the gradient tape hence metric tensor needs to be fed to the optimizer which requires a custom model training function. That's why I believe if the goal is to have maximally flexible software, low-level would be the way to go.

> Regarding the quantum gradient descent optimizer specifically, I'm curious about your implementation, as the original paper took very much a fault-tolerant/non-NISQy approach?

My implementation is mostly based on your implementation here:

https://github.com/PennyLaneAI/pennylane/blob/1fdbcb56e0695668f4e2987550e5999e06fbf640/pennylane/optimize/qng.py#L278-L293

Which is, I believe, mostly based on [your paper](https://quantum-journal.org/papers/q-2020-05-25-269/). So there is no magic, the only problem is TensorFlow has some issues with their linear algorithm package convergence (mainly had this problem while training [MPS via DMRG due to SVD](https://arxiv.org/abs/2106.08334) although it's less likely with `float64` it still struggles to solve linear systems compared to NumPy's implementation) so I have layers of try-except and at the end, if even NumPy fails It reduces the algorithm to regular gradient descent. I realised that especially in large circuits e.g. MERA-based implementations, even NumPy cant find a numerical solution to the linear system; hence the gradient descent.

**Comment by jackaraz** [comment]
> That sounds great @jackaraz! Since it is more low-level, would you be interested in contributing it to via a PR?

Sure, will look into it and create a PR draft. Probably will need to ask for more info about pennylane's functionality to be able to implement it.

**Comment by jackaraz** [comment]
Hi @josh146 I was thinking of writing a Keras optimizer, as shown below.

```python
class QGradientDescent(tf.optimizers.Optimizer):
    def __init__(self, lr: float = 1e-2, **kwargs):
        super(QGradientDescent, self).__init__(**kwargs, name = "QGradientDescent")
        self._set_hyper('learning_rate', lr)


    def apply_gradients(self, grads_vars_metrics: Iterable):
        for grad, var, metric in grads_vars_metrics:
            lr = self._get_hyper("learning_rate", var.dtype)
            try:
                update = lr * tf.linalg.diag_part(
                    tf.linalg.solve(metric, tf.linalg.diag(grad))
                )
            except Exception:
                try:
                    update = tf.convert_to_tensor(
                        lr.numpy() * np.linalg.solve(metric.numpy(), grad.numpy()),
                        dtype = var.dtype
                    )
                except Exception as err:
                    print(
                        "   * failed linalg.solve in both TensorFlow and NumPy "
                        "reducing to gradient descent"
                    )
                    update = lr * grad

            var.assign(var - update)
```
This is like standard Keras optimizer; the only difference is `apply_gradients` takes `zip(grads, trainable_variables, metric_tensors)` instead of `zip(grads, trainable_variables)`, which allows for QGD. But in principle, a model does not necessarily have only quantum layers. It can have a hybrid structure, in which case we won't have the metric tensor, and since the goal is not creating a separate model class, the `apply_gradients` function needs to accept only gradients and trainable parameters. So I thought that the optimizer can be initialized via the model itself. It can identify the quantum layers in the background, which will be required to be pennylane's `KerasLayer`. Then the metric tensor of each quantum layer can be calculated during `apply_gradients` execution, so all classical layers will be trained via SGD, and quantum layers will be trained by QNGD. But this might not be desirable. Another option is to have another optimizer as an input for this class so that classical layers can be trained by that which might affect the efficiency. What do you think would be the best option for pennylane? Do you have much lower lever implementation in mind? I thought modifying the gradient functionality might be an option, but at the end of the day, Keras has a specific structure; if we don't want to go out of traditional Keras usage, there is not much alternative.

**Comment by jackaraz** [comment]
Hi @ixfoduap, thanks for the comments. Yes, I just wanted to make sure that the implementation will be useful for everyone rather than having a very specialized class.

> Just to double-check: have you taken a look at the [quantum natural gradient implementation](https://pennylane.readthedocs.io/en/stable/_modules/pennylane/optimize/qng.html#QNGOptimizer) in PennyLane? I understand that you are addressing specific issues with TF/Keras, but there may be insights from this code that already help you (e.g., I see that metric_tensor_fn=None is passed as a default argument)

Thanks, yes the function above is pretty much a simplified version of the translation for discussion purposes, of course, it requires a bit more control flow.

> There may be an even better alternative I'm missing, but your suggestion of an approach where the classical layers are trained via SGD and the quantum layers are trained with your optimizer sounds reasonable. I'm personally not a huge fan of having a separate optimizer as input for the class.

I totally agree with you giving a separate optimizer will also create the issue of writing a proper scheduler as well so it's better to reduce it to SGD for classical layers, I'm just worried that this will limit the usage of the machinery in a hybrid setting. But if you say its fine, that's good enough for me.

cheers
